ocaml/xapi/xapi_pool.ml

```diff
@@ -626,6 +635,49 @@ let create_or_get_pif_on_master __context rpc session_id (pif_ref, pif) : API.re
 
   new_pif_ref
 
+let create_or_get_vlan_on_master __context rpc session_id (vlan_ref, vlan) : API.ref_VLAN =
+  (* Create a VLAN PIF record only if it is a management PIF *)
+
+  let my_host_ref = Db.PIF.get_host ~__context ~self:vlan.API.vLAN_tagged_PIF in
+  let my_host = Db.Host.get_record ~__context ~self:my_host_ref in
+  let new_host_ref = create_or_get_host_on_master __context rpc session_id (my_host_ref, my_host) in
+
+  (* Create the untagged PIF record on Pool *)
+  let untagged_pif_record = Db.PIF.get_record ~__context ~self:vlan.API.vLAN_untagged_PIF in
+  let remote_untagged_pif = create_or_get_pif_on_master __context rpc session_id
+    (vlan.API.vLAN_untagged_PIF, untagged_pif_record) in
+
+  (* Get the remote tagged pif network *)
+  let tagged_pif_network = Db.PIF.get_network ~__context ~self:vlan.API.vLAN_tagged_PIF in
+  let tagged_pif_network_record = Db.Network.get_record ~__context ~self:tagged_pif_network in
+  let remote_tagged_pif_network = create_or_get_network_on_master __context rpc session_id
+    (tagged_pif_network, tagged_pif_network_record) in
+
+  (* Get the new physical PIF ref on Pool for the joining Host *)
+  let expr =
+    Printf.sprintf "field \"network\"=\"%s\" and field \"host\"=\"%s\""
+    (Ref.string_of remote_tagged_pif_network)
+    (Ref.string_of new_host_ref)
+  in
+  let remote_physical_pif =
```

Comment:
The additional `create_or_get_pif_on_master` above should give you this already.